internal/cephfs/core/metadata.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	fsAdmin "github.com/ceph/go-ceph/cephfs/admin"
+)
+
+const (
+	// clusterNameKey cluster Key, set on RBD image.
+	clusterNameKey = "csi.ceph.com/cluster/name"
+)
+
+// GetMetadata gets custom metadata on the subvolume in a volume based on key
+// name.
+func (s *subVolumeClient) GetMetadata(key string) (string, error) {
+	// keyPresent checks whether corresponding clusterID key is present in clusterAdditionalInfo
+	var keyPresent bool
+	// verify if corresponding clusterID key is present in the map,
+	// and if not, initialize with default values(false).
+	if _, keyPresent = clusterAdditionalInfo[s.clusterID]; !keyPresent {
+		clusterAdditionalInfo[s.clusterID] = &localClusterState{}
+	}
+	// resize subvolume when either it's supported, or when corresponding
+	// clusterID key was not present.
+	if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unknown ||
+		clusterAdditionalInfo[s.clusterID].subVolMetadataState == supported {
+		fsa, err := s.conn.GetFSAdmin()
+		if err != nil {
+			return "", err
+		}
+
+		res, err := fsa.GetMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key)
+		if err == nil {
+			clusterAdditionalInfo[s.clusterID].subVolMetadataState = supported
+
+			return res, nil
+		}
+		var invalid fsAdmin.NotImplementedError
+		// In case the error is other than invalid command return error to the caller.
+		if !errors.As(err, &invalid) {
+			return "", err
+		}
+	}
+	clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported
+
+	return "", fmt.Errorf("operation GetMetadata is not supported")
```

Comment:
```suggestion
notSupErr:=errors.New("operation GetMetadata is not supported")
if  clusterAdditionalInfo[s.clusterID].subVolMetadataState== unsupported {
	return "", notSupErr
}
fsa, err := s.conn.GetFSAdmin()
if err != nil {
	return "", err
}
res, err := fsa.GetMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key)
if err != nil {
  var invalid fsAdmin.NotImplementedError
 // In case the error is other than invalid command return error to the caller.
  if errors.As(err, &invalid) {
  	clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported
  	  err=notSupErr
  }
  return "",err
  }
clusterAdditionalInfo[s.clusterID].subVolMetadataState = supported
return res, nil
}
```

Above suggestions makes it more readable?

Reply:
surely does.